File: packages/agent-sdk/__tests__/protocol-fixtures.test.ts
/**
 * Cross-language protocol fixture contract (TypeScript side).
 *
 * Drives the VS Code extension's real event mapper
 * (`packages/vscode-extension/src/chat/chat-events.js` — the TS/JS panel
 * consumer of Agent Protocol v1) over the SHARED NDJSON fixtures in
 * `__fixtures__/protocol/`, asserting the stable UI projection matches
 * `expected.json`. The JetBrains twin (`ProtocolFixturesTest.java`) reads the
 * SAME files and asserts the same projections against `ChatEvents.java`, so the
 * two IDE panels can never silently diverge.
 *
 * The fixtures are the contract behind docs/PROTOCOL.md — see
 * __fixtures__/protocol/README.md for the projection schema + turn-state rule.
 */
import { createRequire } from "node:module";
import { readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

import { describe, expect, it } from "vitest";

import {
  PROTOCOL_FEATURES,
  isSlashCommandResult,
  type AgentInputEvent,
  type AgentStreamEvent,
  type SlashCommandInput,
  type SlashCommandResultEvent,
  type SystemInitEvent,
  type ToolResultEvent,
} from "../src/protocol.js";

const require = createRequire(import.meta.url);
const here = dirname(fileURLToPath(import.meta.url));
const fixturesDir = join(here, "..", "__fixtures__", "protocol");

// The real panel mapper (CommonJS) — never a re-implementation.
const chatEvents = require(
  join(here, "..", "..", "vscode-extension", "src", "chat", "chat-events.js"),
) as {
  mapAgentEvent: (
    evt: unknown,
    state: { sawDelta: boolean },
  ) => Record<string, unknown> | null;
  createTurnState: () => { sawDelta: boolean };
};

/**
 * Stable projection of a UI message map (README §Projection). Keeps only the
 * machine-comparable dimensions both panels produce byte-identically — never
 * the benign-note wording that legitimately differs between them.
 */
function project(ui: Record<string, unknown> | null): Record<string, unknown> {
  if (ui == null) return { kind: null };
  const kind = ui.kind as string;
  switch (kind) {
    case "init":
      return {
        kind,
        model: ui.model,
        provider: ui.provider,
        sessionId: ui.sessionId,
      };
    case "delta":
    case "thinking":
      return { kind, text: ui.text };
    case "tool":
      return { kind, tool: ui.tool, summary: ui.summary };
    case "tool_done":
      return {
        kind,
        tool: ui.tool,
        isError: ui.isError === true,
        hasNote: ui.note != null,
      };
    case "turn_end":
      return {
        kind,
        isError: ui.isError === true,
        text: ui.text ?? null,
        hasUsage: ui.usage != null,
      };
    case "approval":
      return {
        kind,
        id: ui.id,
        tool: ui.tool ?? null,
        command: ui.command ?? null,
        risk: ui.risk ?? null,
        rule: ui.rule ?? null,
        reason: ui.reason ?? null,
        binding: ui.binding ?? null,
      };
    case "approval_done":
      return { kind, id: ui.id, approved: ui.approved === true, via: ui.via };
    case "question":
      return {
        kind,
        id: ui.id,
        question: ui.question,
        multiSelect: ui.multiSelect === true,
        hasOptions: ui.options != null,
        elicitation: ui.elicitation === true,
        server: ui.server ?? null,
        hasSchema: ui.requestedSchema != null,
        ...(ui.elicitation === true
          ? {
              mode: ui.mode ?? "form",
              elicitationId: ui.elicitationId ?? null,
              url: ui.url ?? null,
              urlHost: ui.urlHost ?? null,
            }
          : {}),
        ...(ui.binding && typeof ui.binding === "object"
          ? { binding: ui.binding }
          : {}),
      };
    case "plan":
      return { kind, active: ui.active === true, state: ui.state ?? null };
    case "usage":
      return { kind };
    case "info":
    case "error":
      return { kind, text: ui.text };
    default:
      return { kind };
  }
}

function readFixtureLines(name: string): unknown[] {
  return readFileSync(join(fixturesDir, name), "utf8")
    .split("\n")
    .filter((l) => l.trim().length > 0)
    .map((l) => JSON.parse(l));
}

const expected = JSON.parse(
  readFileSync(join(fixturesDir, "expected.json"), "utf8"),
) as Record<string, Array<Record<string, unknown>>>;

const causalFixture = JSON.parse(
  readFileSync(join(fixturesDir, "causal-conformance.json"), "utf8"),
) as {
  cases: Array<{ name: string; events: unknown[] }>;
  expected: {
    approvalBinding: { id: string; binding: string };
    terminal: { result: string; isError: boolean };
  };
};

const FIXTURE_FILES = [
  "session-lifecycle.ndjson",
  "assistant-stream.ndjson",
  "tools.ndjson",
  "interaction.ndjson",
  "misc.ndjson",
];

describe("protocol fixture contract (VS Code chat-events twin)", () => {
  it("expected.json covers every fixture file", () => {
    for (const f of FIXTURE_FILES) {
      expect(Array.isArray(expected[f]), `missing expected for ${f}`).toBe(
        true,
      );
    }
  });

  for (const file of FIXTURE_FILES) {
    it(`maps ${file} to the expected UI projections`, () => {
      const events = readFixtureLines(file);
      const want = expected[file];
      expect(events.length, `line count for ${file}`).toBe(want.length);
      // ONE fresh turn-state per file, events fed top-to-bottom (README rule).
      const state = chatEvents.createTurnState();
      const got = events.map((evt) =>
        project(chatEvents.mapAgentEvent(evt, state)),
      );
      expect(got).toEqual(want);
    });
  }

  it("ignores unknown event types (forward compatibility)", () => {
    const state = chatEvents.createTurnState();
    expect(
      chatEvents.mapAgentEvent({ type: "totally_new_event_v9" }, state),
    ).toBe(null);
  });

  it("tolerates the additive seq / trace_id / tool-call id fields without changing mapping", () => {
    const state = chatEvents.createTurnState();
    const withMeta = chatEvents.mapAgentEvent(
      {
        type: "tool_use",
        id: "tu-9",
        tool: "read_file",
        args: { path: "x" },
        seq: 7,
        trace_id: "tr-abc",
      },
      state,
    );
    const withoutMeta = chatEvents.mapAgentEvent(
      { type: "tool_use", tool: "read_file", args: { path: "x" } },
      state,
    );
    expect(withMeta).toEqual(withoutMeta);
  });

  it("preserves causal-equivalent IDE projections across legal interleavings", () => {
    let baseline: string[] | null = null;
    for (const fixtureCase of causalFixture.cases) {
      const state = chatEvents.createTurnState();
      const projections = fixtureCase.events.map((event) =>
        project(chatEvents.mapAgentEvent(event, state)),
      );
      const normalized = projections
        .map((value) => JSON.stringify(value))
        .sort();
      baseline ??= normalized;
      expect(normalized, fixtureCase.name).toEqual(baseline);

      expect(
        projections.find((value) => value.kind === "approval"),
      ).toMatchObject({
        kind: "approval",
        id: causalFixture.expected.approvalBinding.id,
        binding: causalFixture.expected.approvalBinding.binding,
      });
      expect(
        projections.find((value) => value.kind === "turn_end"),
      ).toMatchObject({
        kind: "turn_end",
        isError: causalFixture.expected.terminal.isError,
        text: causalFixture.expected.terminal.result,
        hasUsage: true,
      });
    }
  });
});

describe("session slash-command wire fixture", () => {
  it("types the init capability advertisement", () => {
    const [initValue] = readFixtureLines("session-lifecycle.ndjson");
    const init: SystemInitEvent = initValue as SystemInitEvent;
    const output: AgentStreamEvent = init;

    expect(output.type).toBe("system");
    expect(init.slash_commands).toEqual([
      "status",
      "doctor",
      "mcp",
      "hooks",
      "permissions",
      "agents",
      "tasks",
      "memory",
    ]);
  });

  it("correlates success and failure responses without creating turn results", () => {
    const lines = readFixtureLines("session-slash-command.ndjson");
    expect(lines).toHaveLength(4);

    const [
      statusInputValue,
      statusResultValue,
      deniedInputValue,
      deniedResultValue,
    ] = lines;
    const statusInput = statusInputValue as SlashCommandInput;
    const statusResult = statusResultValue as SlashCommandResultEvent;
    const deniedInput = deniedInputValue as SlashCommandInput;
    const deniedResult = deniedResultValue as SlashCommandResultEvent;

    // Compile-time union coverage: both additive shapes remain accepted by the
    // public AgentInputEvent / AgentStreamEvent contracts.
    const inputs: AgentInputEvent[] = [statusInput, deniedInput];
    const outputs: AgentStreamEvent[] = [statusResult, deniedResult];

    for (const input of inputs) {
      expect(input.type).toBe("slash_command");
      expect(typeof input.request_id).toBe("string");
      expect(typeof input.command).toBe("string");
      expect(typeof input.args).toBe("string");
    }
    for (const output of outputs) {
      expect(isSlashCommandResult(output)).toBe(true);
      expect(output.type).toBe("slash_command_result");
      expect(output.type).not.toBe("result");
    }

    expect(statusResult).toMatchObject({
      request_id: statusInput.request_id,
      command: statusInput.command,
      ok: true,
      text: "Session status",
      session_id: "sess-fx-1",
    });
    expect(statusResult.error).toBeUndefined();

    expect(deniedResult).toMatchObject({
      request_id: deniedInput.request_id,
      command: deniedInput.command,
      ok: false,
      error: {
        code: "UNSUPPORTED_ARGUMENTS",
        message: "/permissions is read-only over stream-json",
      },
      session_id: "sess-fx-1",
    });
    expect(deniedResult.text).toBeUndefined();
  });
});

describe("permission decision wire shape", () => {
  it("advertises the permission decision capability", () => {
    expect(PROTOCOL_FEATURES).toEqual([
      "event_seq",
      "tool_use_id",
      "permission_decision",
      "trace_id",
    ]);
  });

  it("types the additive runtime-authoritative decision on tool_result", () => {
    const event: ToolResultEvent = {
      type: "tool_result",
      id: "tu-1",
      tool: "run_shell",
      is_error: true,
      permission_decision_id: "tu-1:perm:managed",
      permission_decision: {
        version: 1,
        id: "tu-1:perm:managed",
        tool: "run_shell",
        decision: "deny",
        via: "managed",
        rule: "Bash(publish:*)",
        reason: "publishing is disabled",
        chain: [],
      },
    };

    expect(event.permission_decision?.decision).toBe("deny");
    expect(event.permission_decision_id).toBe(event.permission_decision?.id);
  });
});

File: packages/jetbrains-plugin/src/test/java/com/chainlesschain/ide/ProtocolFixturesTest.java
package com.chainlesschain.ide;

import org.junit.jupiter.api.DynamicTest;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestFactory;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

/**
 * Cross-language protocol fixture contract (JetBrains side).
 *
 * <p>Drives the plugin's real event mapper ({@link ChatEvents#mapAgentEvent})
 * over the SAME shared NDJSON fixtures the TypeScript side uses
 * ({@code packages/agent-sdk/__fixtures__/protocol/}, consumed by
 * {@code protocol-fixtures.test.ts} against the VS Code panel's
 * {@code chat-events.js}). Both panels MUST produce byte-identical UI
 * projections on these files — the fixtures are read directly from the
 * agent-sdk package (never copied, never modified), and {@link
 * #sharedFixturesExistAndAreWellFormed()} fails loudly on drift.
 *
 * <p>The contract is docs/PROTOCOL.md — see the fixtures' README.md for the
 * projection schema and the per-file turn-state rule.
 */
class ProtocolFixturesTest {

    private static final String[] FIXTURE_FILES = {
            "session-lifecycle.ndjson", "assistant-stream.ndjson",
            "tools.ndjson", "interaction.ndjson", "misc.ndjson",
    };

    /** Gradle's test working dir is the plugin dir; tolerate the repo root too. */
    private static Path fixturesDir() {
        for (String root : new String[] {
                "../agent-sdk", "packages/agent-sdk",
                "../../packages/agent-sdk" }) {
            Path p = Paths.get(root, "__fixtures__", "protocol");
            if (Files.isDirectory(p)) return p;
        }
        throw new AssertionError(
                "shared protocol twin fixtures not found (fixture drift or moved dir?)"
                        + " cwd=" + Paths.get("").toAbsolutePath());
    }

    private static List<Map<String, Object>> readNdjson(String name) {
        List<Map<String, Object>> out = new ArrayList<>();
        try {
            for (String line : Files.readAllLines(
                    fixturesDir().resolve(name), StandardCharsets.UTF_8)) {
                if (line.trim().isEmpty()) continue;
                out.add(MiniJson.parseObject(line));
            }
        } catch (IOException e) {
            throw new AssertionError("cannot read fixture " + name, e);
        }
        return out;
    }

    @SuppressWarnings("unchecked")
    private static Map<String, Object> expectedDoc() {
        try {
            return (Map<String, Object>) MiniJson.parse(Files.readString(
                    fixturesDir().resolve("expected.json"), StandardCharsets.UTF_8));
        } catch (IOException e) {
            throw new AssertionError("cannot read expected.json", e);
        }
    }

    @SuppressWarnings("unchecked")
    private static List<Object> expectedFor(String file) {
        Object v = expectedDoc().get(file);
        assertTrue(v instanceof List, "missing expected for " + file);
        return (List<Object>) v;
    }

    private static Object orNull(Map<String, Object> ui, String key) {
        return ui == null ? null : ui.get(key);
    }

    private static boolean truthy(Object v) {
        return Boolean.TRUE.equals(v);
    }

    /**
     * Stable projection of a UI message map (README §Projection) — the exact
     * Java twin of the TS test's {@code project()}. Keeps only the
     * machine-comparable dimensions; the benign {@code tool_done} note wording
     * (which differs between the panels by design) collapses to {@code hasNote}.
     */
    private static Map<String, Object> project(Map<String, Object> ui) {
        Map<String, Object> p = new LinkedHashMap<>();
        if (ui == null) {
            p.put("kind", null);
            return p;
        }
        String kind = String.valueOf(ui.get("kind"));
        p.put("kind", kind);
        switch (kind) {
            case "init":
                p.put("model", ui.get("model"));
                p.put("provider", ui.get("provider"));
                p.put("sessionId", ui.get("sessionId"));
                break;
            case "delta":
            case "thinking":
                p.put("text", ui.get("text"));
                break;
            case "tool":
                p.put("tool", ui.get("tool"));
                p.put("summary", ui.get("summary"));
                break;
            case "tool_done":
                p.put("tool", ui.get("tool"));
                p.put("isError", truthy(ui.get("isError")));
                p.put("hasNote", ui.get("note") != null);
                break;
            case "turn_end":
                p.put("isError", truthy(ui.get("isError")));
                p.put("text", ui.get("text"));
                p.put("hasUsage", ui.get("usage") != null);
                break;
            case "approval":
                p.put("id", ui.get("id"));
                p.put("tool", orNull(ui, "tool"));
                p.put("command", orNull(ui, "command"));
                p.put("risk", orNull(ui, "risk"));
                p.put("rule", orNull(ui, "rule"));
                p.put("reason", orNull(ui, "reason"));
                break;
            case "approval_done":
                p.put("id", ui.get("id"));
                p.put("approved", truthy(ui.get("approved")));
                p.put("via", ui.get("via"));
                break;
            case "question":
                p.put("id", ui.get("id"));
                p.put("question", ui.get("question"));
                p.put("multiSelect", truthy(ui.get("multiSelect")));
                p.put("hasOptions", ui.get("options") != null);
                p.put("elicitation", truthy(ui.get("elicitation")));
                p.put("server", ui.get("server"));
                p.put("hasSchema", ui.get("requestedSchema") != null);
                break;
            case "plan":
                p.put("active", truthy(ui.get("active")));
                p.put("state", ui.get("state"));
                break;
            case "usage":
                break;
            case "info":
            case "error":
                p.put("text", ui.get("text"));
                break;
            default:
                break;
        }
        return p;
    }

    @Test
    void sharedFixturesExistAndAreWellFormed() {
        // Drift guard: the twin contract lives in the agent-sdk package; a
        // moved/renamed fixture must fail HERE, not silently run zero cases.
        Path dir = fixturesDir();
        for (String f : FIXTURE_FILES) {
            assertTrue(Files.isRegularFile(dir.resolve(f)), "missing fixture " + f);
        }
        assertTrue(Files.isRegularFile(dir.resolve("expected.json")),
                "missing expected.json");
        Map<String, Object> exp = expectedDoc();
        for (String f : FIXTURE_FILES) {
            assertTrue(exp.get(f) instanceof List, "missing expected for " + f);
        }
    }

    @TestFactory
    List<DynamicTest> mapsEachFixtureToExpectedProjections() {
        List<DynamicTest> tests = new ArrayList<>();
        for (String file : FIXTURE_FILES) {
            tests.add(DynamicTest.dynamicTest("maps " + file, () -> {
                List<Map<String, Object>> events = readNdjson(file);
                List<Object> want = expectedFor(file);
                assertEquals(want.size(), events.size(), "line count for " + file);
                // ONE fresh turn-state per file, events fed top-to-bottom.
                ChatEvents.TurnState state = new ChatEvents.TurnState();
                for (int i = 0; i < events.size(); i++) {
                    Map<String, Object> got = project(
                            ChatEvents.mapAgentEvent(events.get(i), state));
                    assertEquals(want.get(i), got,
                            file + " line " + (i + 1));
                }
            }));
        }
        return tests;
    }

    @Test
    void ignoresUnknownEventTypes() {
        ChatEvents.TurnState state = new ChatEvents.TurnState();
        Map<String, Object> evt = new LinkedHashMap<>();
        evt.put("type", "totally_new_event_v9");
        assertNull(ChatEvents.mapAgentEvent(evt, state));
    }

    @Test
    void toleratesAdditiveSeqAndToolIdFields() {
        ChatEvents.TurnState state = new ChatEvents.TurnState();
        Map<String, Object> withMeta = new LinkedHashMap<>();
        withMeta.put("type", "tool_use");
        withMeta.put("id", "tu-9");
        withMeta.put("tool", "read_file");
        withMeta.put("seq", 7L);
        Map<String, Object> args = new LinkedHashMap<>();
        args.put("path", "x");
        withMeta.put("args", args);

        Map<String, Object> withoutMeta = new LinkedHashMap<>();
        withoutMeta.put("type", "tool_use");
        withoutMeta.put("tool", "read_file");
        Map<String, Object> args2 = new LinkedHashMap<>();
        args2.put("path", "x");
        withoutMeta.put("args", args2);

        assertEquals(
                ChatEvents.mapAgentEvent(withoutMeta, state),
                ChatEvents.mapAgentEvent(withMeta, state));
    }
}
